### PR TITLE
fix #6954

### LINKED ISSUE #6954: getstr in bin_dex.c returns NULL
which is sometimes used in a format string or concatenated which may produce a null deref (not all libcs handle null pointers in the way that gnu libc does)

### PR DESCRIPTION
**Pull request by h4ng3r**
getstr know returns a pointer to string from the binary. If its invalid it returns "" instead of NULL, so it fixes the deref.
I reduced the memory operations performed in getstr, now only half of the strings use heap. I still using heap cuz "r_str_newf" and "r_str_replace".

### CODE REVIEW

File: libr/bin/p/bin_dex.c [hunk 1]
```diff
@@ -368,7 +353,7 @@ static void dex_parse_debug_item(RBinFile *binfile, RBinDexObj *bin,
 			argReg += 1;
 			break;
 		}
-		if (name) {
+		if (name != "") {
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: if (*name)

File: libr/bin/p/bin_dex.c [hunk 2]
```diff
@@ -839,18 +824,18 @@ static char *dex_class_name(RBinDexObj *bin, RBinDexClass *c) {
 static char *dex_field_name(RBinDexObj *bin, int fid) {
 	int cid, tid, type_id;
 	if (!bin || !bin->fields) {
-		return NULL;
+		return "";
 	}
 	if (fid < 0 || fid >= bin->header.fields_size) {
-		return NULL;
+		return strdup ("");
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Return "", without strdup

File: libr/bin/p/bin_dex.c [hunk 3]
```diff
@@ -859,21 +844,21 @@ static char *dex_field_name(RBinDexObj *bin, int fid) {
...
 	}
 	int cid = bin->methods[method_idx].class_id;
 	if (cid < 0 || cid >= bin->header.types_size) {
-		return NULL;
+		return strdup ("");
 	}
 	char *name = dex_method_name (bin, method_idx);
-	char *class_name = dex_class_name_byid (bin, cid);
+	char *class_name = strdup (dex_class_name_byid (bin, cid));
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: No need to strdup

File: libr/bin/p/bin_dex.c [hunk 4]
```diff
@@ -1280,8 +1265,11 @@ static void parse_class(RBinFile *binfile, RBinDexObj *bin, RBinDexClass *c,
 	if (!c) {
 		return;
 	}
-
 	class_name = dex_class_name (bin, c);
+	if (class_name == "") {
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: Chk if (*class_name

File: libr/bin/p/bin_dex.c (hunk 3 above)
Comment (h4ng3r) [inline libr/bin/p/bin_dex.c]: Without the strdup it crashes, because "r_str_replace" does a free on all params.

**Pull request closed without merging**